assert/assertions.go

```diff
@@ -522,29 +528,37 @@ func NotSame(t TestingT, expected, actual interface{}, msgAndArgs ...interface{}
 		h.Helper()
 	}
 
-	if samePointers(expected, actual) {
+	same, ok := samePointers(expected, actual)
+	if !ok {
+		//fails when the arguments are not pointers
+		return !(Fail(t, "Both arguments must be pointers", msgAndArgs...))
+	}
+
+	if same {
 		return Fail(t, fmt.Sprintf(
 			"Expected and actual point to the same object: %p %#v",
 			expected, expected), msgAndArgs...)
 	}
 	return true
 }
 
-// samePointers compares two generic interface objects and returns whether
-// they point to the same object
-func samePointers(first, second interface{}) bool {
+// samePointers checks if two generic interface objects are pointers to
+// the same object. It returns two values: a boolean indicating if they
+// point to the same object, and another boolean indicating whether both
+// inputs are valid pointers of the same type.
+func samePointers(first, second interface{}) (same bool, ok bool) {
```

Comment:
Sorry, I should have noticed this on the last pass. The type comparison is yielded in `same` and not in `ok`.

```suggestion
// samePointers checks if two generic interface objects are pointers of the same
// type pointing to the same object. It returns two values: same indicating if
// they are the same type and point to the same object, and ok indicating that
// both inputs are pointers.
func samePointers(first, second interface{}) (same bool, ok bool) {
```

Reply:
I've made the change :) Let me know if any other changes are to be made